sycl/include/sycl/ext/oneapi/group_algorithm.hpp

```diff
@@ -65,6 +65,125 @@ using EnableIfIsNonNativeOp = cl::sycl::detail::enable_if_t<
+  if constexpr (std::is_same<Group, sycl::ext::oneapi::sub_group>::value) {
+    return __spv::Scope::Subgroup;
+  } else {
+    return __spv::Scope::Workgroup;
+  }
+}
+} // namespace detail
+
+/// Asynchronously copies a number of elements specified by \p numElements
+/// from the source pointed by \p src to destination pointed by \p dest
+/// with a source stride specified by \p srcStride, and returns a SYCL
+/// device_event which can be used to wait on the completion of the copy.
+/// Permitted types for dataT are all scalar and vector types, except boolean.
+template <typename Group, typename dataT>
+detail::enable_if_t<is_group_v<Group> && !detail::is_bool<dataT>::value,
+                    device_event>
+async_group_copy(Group, global_ptr<dataT> src, local_ptr<dataT> dest,
+                 size_t numElements, size_t srcStride) {
+  using DestT = detail::ConvertToOpenCLType_t<decltype(dest)>;
+  using SrcT = detail::ConvertToOpenCLType_t<decltype(src)>;
+
+  __ocl_event_t E = __SYCL_OpGroupAsyncCopyGlobalToLocal(
+      detail::group_to_scope<Group>(), DestT(dest.get()), SrcT(src.get()),
+      numElements, srcStride, 0);
+  return device_event(&E);
+}
+
+/// Asynchronously copies a number of elements specified by \p numElements
+/// from the source pointed by \p src to destination pointed by \p dest with
+/// the destination stride specified by \p destStride, and returns a SYCL
+/// device_event which can be used to wait on the completion of the copy.
+/// Permitted types for dataT are all scalar and vector types, except boolean.
+template <typename Group, typename dataT>
+detail::enable_if_t<is_group_v<Group> && !detail::is_bool<dataT>::value,
+                    device_event>
+async_group_copy(Group, local_ptr<dataT> src, global_ptr<dataT> dest,
+                 size_t numElements, size_t destStride) {
+  using DestT = detail::ConvertToOpenCLType_t<decltype(dest)>;
+  using SrcT = detail::ConvertToOpenCLType_t<decltype(src)>;
+
+  __ocl_event_t E = __SYCL_OpGroupAsyncCopyLocalToGlobal(
+      detail::group_to_scope<Group>(), DestT(dest.get()), SrcT(src.get()),
+      numElements, destStride, 0);
+  return device_event(&E);
+}
+
+/// Specialization for scalar bool type.
+/// Asynchronously copies a number of elements specified by \p NumElements
+/// from the source pointed by \p Src to destination pointed by \p Dest
+/// with a stride specified by \p Stride, and returns a SYCL device_event
+/// which can be used to wait on the completion of the copy.
+template <typename Group, typename T, access::address_space DestS,
+          access::address_space SrcS>
+detail::enable_if_t<is_group_v<Group> && detail::is_scalar_bool<T>::value,
+                    device_event>
+async_group_copy(Group g, multi_ptr<T, SrcS> Src, multi_ptr<T, DestS> Dest,
+                 size_t NumElements, size_t Stride) {
+  static_assert(sizeof(bool) == sizeof(uint8_t),
+                "Async copy to/from bool memory is not supported.");
```

Comment:
I guess the idea here is that `bool` takes 1 byte? Wouldn't it be better then to replace `uint8_t` with `char`?

Reply:
Yes, I think that is the idea. I've replaced that now.